src/Language/SystemVerilog/Parser/Preprocess.hs

```diff
@@ -661,36 +694,21 @@ handleDirective macrosOnly = do
         "ifdef" -> do
             dropSpaces
             name <- takeIdentifier
-            let newCond = if Map.member name env
-                    then CurrentlyTrue
-                    else NeverTrue
-            setCondStack $ newCond : condStack
+            pushCondStack $ boolToCond $ Map.member name env
         "ifndef" -> do
             dropSpaces
             name <- takeIdentifier
-            let newCond = if Map.notMember name env
-                    then CurrentlyTrue
-                    else NeverTrue
-            setCondStack $ newCond : condStack
+            pushCondStack $ elseCond $ boolToCond $ Map.member name env
         "else" -> do
-            let newCond = if head condStack == NeverTrue
-                    then CurrentlyTrue
-                    else NeverTrue
-            setCondStack $ newCond : tail condStack
+            c <- popCondStack "else"
+            pushCondStack $ elseCond c
         "elsif" -> do
             dropSpaces
             name <- takeIdentifier
-            let currCond = head condStack
-            let newCond =
-                    if currCond /= NeverTrue then
-                        PreviouslyTrue
-                    else if Map.member name env then
-                        CurrentlyTrue
-                    else
-                        NeverTrue
-            setCondStack $ newCond : tail condStack
+            c <- popCondStack "elsif"
+            pushCondStack $ elsifCond (Map.member name env) c
         "endif" -> do
-            setCondStack $ tail condStack
+            popCondStack "endif" >> return ()
```

Comment:
Oops. Yep, sorted (although you need to explicitly discard the result of popCondStack to avoid a GHC warning, which I suspect is why I wrote the `>>` version in the first place)